[network] Rename Network Role to Network Type

gregnazario opened the pull request:
RoleType usage in Network was so confusing, as it represented the type of network (Validator vs FullNode), which was kinda silly when it was one network Id vs all other network Ids.  This change replaces RoleType with NetworkType for appropriate usages, and puts the actual RoleType of the Node into the NetworkContext for logging and metrics purposes.

gregnazario commented:
/canary

bors-libra commented:
:sunny: Canary successful

davidiw reviewed (commented):
why not just use network id? while a step in the right direction ... it feels like there's an opportunity to reuse rather copy paste.

phlip9 commented on config/src/config/mod.rs:
```diff
@@ -210,6 +210,54 @@ impl fmt::Display for RoleType {
+pub enum NetworkType {
+    Validator,
+    FullNode,
+}
```
is this type a projection on `NetworkId`? in other words, every `NetworkId` is one of these `NetworkType`s? it would make sense to me to add `NetworkId::network_type(&self) -> NetworkType` then.

as for the usage, it does seem useful for distinguishing e.g. a node with role=Validator and network_type=FullNode.

gregnazario replied:
Yeah, it is.  I can come back around and just remove the `NetworkType` mostly.  There's just some places where having a type is easier than dealing with converting to an `is_validator`

phlip9 commented on config/src/network_id.rs:
```diff
@@ -147,6 +152,18 @@ impl NetworkId {
+        } else {
+            NetworkType::FullNode
+        }
+    }
```
oh I spoke too soon haha, looks like it's here 

I would probably rename this to just `network_type` though

phlip9 commented on config/src/config/network_config.rs:
```diff
@@ -167,20 +167,28 @@ impl NetworkConfig {
+    }
+
+    pub fn load_fullnode_network(&mut self) -> Result<(), Error> {
+        if self.network_id.is_validator_network() {
```
cool cool, this is basically what I expected. we just have the `network_id` in the config and then project into a `NetworkType` when needed.

phlip9 commented on config/src/config/network_config.rs:
```diff
@@ -167,20 +167,28 @@ impl NetworkConfig {
+    pub fn load_validator_network(&mut self) -> Result<(), Error> {
+        self.network_id = NetworkId::Validator;
+        self.load()
+    }
```
doesn't ignoring the network_id like this kind of defeat the purpose of checking the invariants? i know this is the same behavior as before, but maybe we can make this safer

gregnazario replied:
So, I tried changing this, because I agree with you.

However, this is used as basically a "skip defining the validator network type" method.  So, without changing every validator config, it would crash if I added this here.  It's more about the fact that since we share a config type between the two that makes this rough.

gregnazario commented on config/management/operational/src/network_checker.rs:
```diff
@@ -94,12 +94,12 @@ pub struct CheckValidatorSetEndpoints {
     /// Specifies whether or not to evaluate validators or fullnodes
     #[structopt(long)]
-    role: RoleType,
+    network_type: NetworkType,
```
Yes, I believe I'd have to, maybe I should just name them back to role.

gregnazario commented:
> why not just use network id? while a step in the right direction ... it feels like there's an opportunity to reuse rather copy paste.

There are a couple places that need to still use a separate type, but I can go and work out a few more of them.

gregnazario commented:
> why not just use network id? while a step in the right direction ... it feels like there's an opportunity to reuse rather copy paste.

@davidiw I just reverted the two places that needed it back to RoleType (tools).

The original reason for the copy paste on the RoleType, was I wasn't quite sure how to handle the metrics / logging differences and I had removed `Role` entirely from `NetworkContext`.  Once, the real `RoleType` was reintroduced, it became easy to delete it because without something in that spot, all the dashboards / alarms would have needed to be updated.

gregnazario commented:
/land
